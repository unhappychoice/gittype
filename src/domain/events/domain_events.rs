use super::Event;
use std::any::Any;

// Unified domain event enum for pattern matching
#[derive(Debug, Clone)]
pub enum DomainEvent {
    KeyPressed { key: char, position: usize },
    StageStarted { start_time: std::time::Instant },
    StagePaused,
    StageResumed,
    StageFinalized,
    StageSkipped,
    ChallengeLoaded { text: String, source_path: String },
}

impl Event for DomainEvent {
    fn as_any(&self) -> &dyn Any {
        self
    }
}
