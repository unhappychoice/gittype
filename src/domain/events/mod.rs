use std::any::{Any, TypeId};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

pub mod domain_events;

pub trait Event: Send + Sync + 'static {
    fn as_any(&self) -> &dyn Any;
}

pub trait EventHandler<E: Event>: Send + Sync {
    fn handle(&self, event: &E);
}

type BoxedHandler = Arc<dyn Fn(&dyn Event) + Send + Sync>;

pub struct EventBus {
    subscribers: Arc<RwLock<HashMap<TypeId, Vec<BoxedHandler>>>>,
}

impl EventBus {
    pub fn new() -> Self {
        Self {
            subscribers: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub fn publish<E: Event>(&self, event: E) {
        let type_id = TypeId::of::<E>();

        // Clone Arc<handlers> while holding the lock, then release it before calling them
        let handlers: Vec<BoxedHandler> = {
            let subscribers = self.subscribers.read().unwrap();
            let h = subscribers.get(&type_id)
                .map(|h| h.clone())
                .unwrap_or_default();
            h
        }; // Lock is released here

        // Call handlers without holding the lock
        for handler in handlers.iter() {
            handler(&event);
        }
    }

    pub fn subscribe<E: Event, F>(&self, handler: F)
    where
        F: Fn(&E) + Send + Sync + 'static,
    {
        let type_id = TypeId::of::<E>();
        let mut subscribers = self.subscribers.write().unwrap();

        let boxed_handler: BoxedHandler = Arc::new(move |event: &dyn Event| {
            if let Some(concrete_event) = event.as_any().downcast_ref::<E>() {
                handler(concrete_event);
            }
        });

        subscribers
            .entry(type_id)
            .or_insert_with(Vec::new)
            .push(boxed_handler);
    }
}

impl Default for EventBus {
    fn default() -> Self {
        Self::new()
    }
}

impl Clone for EventBus {
    fn clone(&self) -> Self {
        Self {
            subscribers: Arc::clone(&self.subscribers),
        }
    }
}
