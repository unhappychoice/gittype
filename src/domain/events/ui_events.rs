use super::Event;
use std::any::Any;

#[derive(Debug, Clone)]
pub struct RenderRequested {
    pub screen_name: String,
}

impl Event for RenderRequested {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Debug, Clone)]
pub struct ScreenChanged {
    pub from: String,
    pub to: String,
}

impl Event for ScreenChanged {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Debug, Clone)]
pub struct UserInputReceived {
    pub input: String,
}

impl Event for UserInputReceived {
    fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Debug, Clone)]
pub struct ConfigUpdated {
    pub key: String,
    pub value: String,
}

impl Event for ConfigUpdated {
    fn as_any(&self) -> &dyn Any {
        self
    }
}
