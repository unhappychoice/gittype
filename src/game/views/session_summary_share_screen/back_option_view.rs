use crate::ui::Colors;
use crate::Result;
use crossterm::{
    cursor::MoveTo,
    execute,
    style::{Print, ResetColor, SetForegroundColor},
};
use std::io::{stdout, Write};

pub struct BackOptionView;

impl BackOptionView {
    pub fn render(center_col: u16, row: u16) -> Result<()> {
        let mut stdout = stdout();

        let back_key = "[ESC]";
        let back_label = " Back to Results";
        let full_back_len = back_key.len() + back_label.len();
        let back_col = center_col.saturating_sub(full_back_len as u16 / 2);

        execute!(stdout, MoveTo(back_col, row))?;
        execute!(
            stdout,
            SetForegroundColor(Colors::to_crossterm(Colors::SUCCESS))
        )?;
        execute!(stdout, Print(back_key))?;
        execute!(
            stdout,
            SetForegroundColor(Colors::to_crossterm(Colors::TEXT))
        )?;
        execute!(stdout, Print(back_label))?;
        execute!(stdout, ResetColor)?;

        stdout.flush()?;
        Ok(())
    }
}
