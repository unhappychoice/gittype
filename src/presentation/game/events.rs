use std::any::Any;

// Re-export ScreenTransition as NavigateTo event
pub use super::models::screen::ScreenTransition as NavigateTo;

/// Event emitted when user requests to exit the application (Ctrl+C)
#[derive(Clone, Debug)]
pub struct ExitRequested;

impl crate::domain::events::Event for ExitRequested {
    fn as_any(&self) -> &dyn Any {
        self
    }
}
