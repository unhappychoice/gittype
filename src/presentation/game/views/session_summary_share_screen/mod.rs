pub mod back_option_view;
pub mod platform_options_view;
pub mod preview_view;
pub mod title_view;

pub use back_option_view::BackOptionView;
pub use platform_options_view::PlatformOptionsView;
pub use preview_view::PreviewView;
pub use title_view::TitleView;
