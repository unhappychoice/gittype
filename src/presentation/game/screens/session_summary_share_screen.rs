use crate::presentation::game::models::{Screen, ScreenTransition, UpdateStrategy};
use crate::presentation::game::views::{
    ShareBackOptionView, SharePlatformOptionsView, SharePreviewView, ShareTitleView,
};
use crate::presentation::sharing::{SharingPlatform, SharingService};
use crate::{domain::models::GitRepository, Result};
use crossterm::terminal::{self};
use std::io::Stdout;

pub struct SessionSummaryShareScreen {
    session_result: Option<crate::domain::models::SessionResult>,
    git_repository: Option<crate::domain::models::GitRepository>,
}

impl Default for SessionSummaryShareScreen {
    fn default() -> Self {
        Self::new()
    }
}

impl SessionSummaryShareScreen {
    pub fn new() -> Self {
        Self {
            session_result: None,
            git_repository: None,
        }
    }

    fn render(
        metrics: &crate::domain::models::SessionResult,
        repo_info: &Option<GitRepository>,
    ) -> Result<()> {
        let (terminal_width, terminal_height) = terminal::size()?;
        let center_row = terminal_height / 2;
        let center_col = terminal_width / 2;

        let platforms = SharingPlatform::all();
        let start_row = center_row.saturating_sub(2);

        ShareTitleView::render(center_col, center_row)?;
        SharePreviewView::render(metrics, repo_info, center_col, center_row)?;
        SharePlatformOptionsView::render(center_col, start_row)?;
        ShareBackOptionView::render(center_col, start_row + platforms.len() as u16 + 2)?;

        Ok(())
    }
}

impl Screen for SessionSummaryShareScreen {
    fn handle_key_event(
        &mut self,
        key_event: crossterm::event::KeyEvent,
    ) -> Result<ScreenTransition> {
        use crossterm::event::{KeyCode, KeyModifiers};
        match key_event.code {
            KeyCode::Char('1') => {
                if let Some(ref session_result) = self.session_result {
                    let _ = SharingService::share_result(
                        session_result,
                        SharingPlatform::X,
                        &self.git_repository,
                    );
                }
                Ok(ScreenTransition::Pop)
            }
            KeyCode::Char('2') => {
                if let Some(ref session_result) = self.session_result {
                    let _ = SharingService::share_result(
                        session_result,
                        SharingPlatform::Reddit,
                        &self.git_repository,
                    );
                }
                Ok(ScreenTransition::Pop)
            }
            KeyCode::Char('3') => {
                if let Some(ref session_result) = self.session_result {
                    let _ = SharingService::share_result(
                        session_result,
                        SharingPlatform::LinkedIn,
                        &self.git_repository,
                    );
                }
                Ok(ScreenTransition::Pop)
            }
            KeyCode::Char('4') => {
                if let Some(ref session_result) = self.session_result {
                    let _ = SharingService::share_result(
                        session_result,
                        SharingPlatform::Facebook,
                        &self.git_repository,
                    );
                }
                Ok(ScreenTransition::Pop)
            }
            KeyCode::Esc => Ok(ScreenTransition::Pop),
            KeyCode::Char('c') if key_event.modifiers.contains(KeyModifiers::CONTROL) => {
                Ok(ScreenTransition::Exit)
            }
            _ => Ok(ScreenTransition::None),
        }
    }

    fn render_crossterm_with_data(
        &mut self,
        _stdout: &mut Stdout,
        session_result: Option<&crate::domain::models::SessionResult>,
        _total_result: Option<&crate::domain::services::scoring::TotalResult>,
    ) -> Result<()> {
        self.session_result = session_result.cloned();
        // Get git repository from global GameData
        let git_repository = crate::presentation::game::game_data::GameData::get_git_repository();
        self.git_repository = git_repository.clone();

        if let Some(session_result) = session_result {
            Self::render(session_result, &git_repository)?;
        }
        Ok(())
    }

    fn get_update_strategy(&self) -> UpdateStrategy {
        UpdateStrategy::InputOnly
    }

    fn update(&mut self) -> Result<bool> {
        Ok(false)
    }

    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }
}
