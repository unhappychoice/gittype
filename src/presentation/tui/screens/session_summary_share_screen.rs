use crate::domain::events::EventBus;
use crate::domain::models::SessionResult;
use crate::presentation::game::events::NavigateTo;
use crate::presentation::game::{GameData, SessionManager};
use crate::presentation::sharing::{SharingPlatform, SharingService};
use crate::presentation::tui::views::{
    ShareBackOptionView, SharePlatformOptionsView, SharePreviewView, ShareTitleView,
};
use crate::presentation::tui::{Screen, ScreenDataProvider, ScreenType, UpdateStrategy};
use crate::{domain::models::GitRepository, GitTypeError, Result};
use ratatui::{
    layout::{Constraint, Direction, Layout},
    Frame,
};
use std::sync::{Arc, Mutex};

pub struct SessionSummaryShareData {
    pub session_result: SessionResult,
    pub git_repository: Option<GitRepository>,
}

pub struct SessionSummaryShareDataProvider {
    session_manager: Arc<Mutex<SessionManager>>,
    game_data: Arc<Mutex<GameData>>,
}

impl ScreenDataProvider for SessionSummaryShareDataProvider {
    fn provide(&self) -> Result<Box<dyn std::any::Any>> {
        let session_result = self
            .session_manager
            .lock()
            .map_err(|_| GitTypeError::TerminalError("Failed to lock SessionManager".to_string()))?
            .get_session_result()
            .ok_or_else(|| {
                GitTypeError::TerminalError("No session result available".to_string())
            })?;

        let git_repository = self
            .game_data
            .lock()
            .map_err(|_| GitTypeError::TerminalError("Failed to lock GameData".to_string()))?
            .git_repository
            .clone();

        Ok(Box::new(SessionSummaryShareData {
            session_result,
            git_repository,
        }))
    }
}

pub struct SessionSummaryShareScreen {
    session_result: Option<SessionResult>,
    git_repository: Option<GitRepository>,
    event_bus: EventBus,
}

impl SessionSummaryShareScreen {
    pub fn new(event_bus: EventBus) -> Self {
        Self {
            session_result: None,
            git_repository: None,
            event_bus,
        }
    }
}

impl Screen for SessionSummaryShareScreen {
    fn get_type(&self) -> ScreenType {
        ScreenType::SessionSharing
    }

    fn default_provider() -> Box<dyn crate::presentation::tui::ScreenDataProvider>
    where
        Self: Sized,
    {
        Box::new(SessionSummaryShareDataProvider {
            session_manager: SessionManager::instance(),
            game_data: GameData::instance(),
        })
    }

    fn init_with_data(&mut self, data: Box<dyn std::any::Any>) -> Result<()> {
        let data = data.downcast::<SessionSummaryShareData>()?;

        self.session_result = Some(data.session_result);
        self.git_repository = data.git_repository;

        Ok(())
    }

    fn handle_key_event(&mut self, key_event: crossterm::event::KeyEvent) -> Result<()> {
        use crossterm::event::{KeyCode, KeyModifiers};
        match key_event.code {
            KeyCode::Char('1') => {
                if let Some(ref session_result) = self.session_result {
                    let _ = SharingService::share_result(
                        session_result,
                        SharingPlatform::X,
                        &self.git_repository,
                    );
                }
                self.event_bus.publish(NavigateTo::Pop);
                Ok(())
            }
            KeyCode::Char('2') => {
                if let Some(ref session_result) = self.session_result {
                    let _ = SharingService::share_result(
                        session_result,
                        SharingPlatform::Reddit,
                        &self.git_repository,
                    );
                }
                self.event_bus.publish(NavigateTo::Pop);
                Ok(())
            }
            KeyCode::Char('3') => {
                if let Some(ref session_result) = self.session_result {
                    let _ = SharingService::share_result(
                        session_result,
                        SharingPlatform::LinkedIn,
                        &self.git_repository,
                    );
                }
                self.event_bus.publish(NavigateTo::Pop);
                Ok(())
            }
            KeyCode::Char('4') => {
                if let Some(ref session_result) = self.session_result {
                    let _ = SharingService::share_result(
                        session_result,
                        SharingPlatform::Facebook,
                        &self.git_repository,
                    );
                }
                self.event_bus.publish(NavigateTo::Pop);
                Ok(())
            }
            KeyCode::Esc => {
                self.event_bus.publish(NavigateTo::Pop);
                Ok(())
            }
            KeyCode::Char('c') if key_event.modifiers.contains(KeyModifiers::CONTROL) => {
                self.event_bus.publish(NavigateTo::Exit);
                Ok(())
            }
            _ => Ok(()),
        }
    }

    fn render_ratatui(&mut self, frame: &mut Frame) -> Result<()> {
        if let Some(ref session_result) = self.session_result {
            let area = frame.area();

            let content_height = 12;
            let top_spacing = (area.height.saturating_sub(content_height)) / 2;

            let chunks = Layout::default()
                .direction(Direction::Vertical)
                .constraints([
                    Constraint::Length(top_spacing),
                    Constraint::Length(1),
                    Constraint::Length(3),
                    Constraint::Length(1),
                    Constraint::Length(1),
                    Constraint::Length(4),
                    Constraint::Length(1),
                    Constraint::Length(1),
                    Constraint::Min(0),
                ])
                .split(area);

            ShareTitleView::render(frame, chunks[1]);
            SharePreviewView::render(frame, chunks[3], session_result, &self.git_repository);
            SharePlatformOptionsView::render(frame, chunks[5]);
            ShareBackOptionView::render(frame, chunks[7]);
        }
        Ok(())
    }

    fn get_update_strategy(&self) -> UpdateStrategy {
        UpdateStrategy::InputOnly
    }

    fn update(&mut self) -> Result<bool> {
        Ok(false)
    }

    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }
}
