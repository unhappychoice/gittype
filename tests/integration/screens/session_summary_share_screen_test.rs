use crate::integration::screens::mocks::session_summary_share_screen_mock::MockSessionSummaryShareDataProvider;
use gittype::domain::events::EventBus;
use gittype::presentation::game::screens::session_summary_share_screen::SessionSummaryShareScreen;

screen_snapshot_test!(
    test_session_summary_share_screen_snapshot,
    SessionSummaryShareScreen,
    SessionSummaryShareScreen::new(EventBus::new()),
    provider = MockSessionSummaryShareDataProvider
);
