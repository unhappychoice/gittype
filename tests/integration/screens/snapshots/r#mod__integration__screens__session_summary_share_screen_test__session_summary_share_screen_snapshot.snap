---
source: tests/integration/screens/session_summary_share_screen_test.rs
expression: output
---
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
                                                === SHARE YOUR RESULT ===                                               
                                                                                                                        
                                                                                                                        
                                                                                                                        
                  "Security Engineer" with 8500pts on [unhappychoice/gittype] - CPM: 225, Mistakes: 35                  
                                                                                                                        
                                                          [1] X                                                         
                                                       [2] Reddit                                                       
                                                      [3] LinkedIn                                                      
                                                      [4] Facebook                                                      
                                                                                                                        
                                                  [ESC] Back to Results
